### DISCUSSION TITLE
Milliseconds in RichHandler timestamps

### DISCUSSION THREAD

**Discussion by finnhughes** [Textualize/rich#984] [Q&A]
Hi, I'm new to this fantastic looking library, thanks.

I want to use RichHandler for logging and like milliseconds in my timestamps. For now I've specified `datefmt='%H:%M:%S.%f'` but I'd prefer not to have microseconds. In the python formatter if you don't specify datefmt you get milliseconds by default but that's not possible with RichHandler. With the python formatter you can embed the timestamp in the message but then RichHandler doesn't recognise it as a timestamp anymore.

I had a quick look to see if there was something I could easily override but my best guess was overriding the LogRender __call__ dunder function and that didn't feel very tidy.

Any thoughts?

**Comment by willmcgugan** [comment]
Rich uses `datetime.strftime` to format timestamps, but it doesn't look like that supports milliseconds (surprisingly). If you can't live without it, file a feature requests. No promises on timescale. Of course, if you think you could tackle it yourself I'd accept a PR...

**Reply by finnhughes** [reply]
Thanks, I've had a quick look and a PR seems easy enough but I'm not sure how you'd like it configured.

I can see some value in emulating the class variables from the Formatter class in logging:
```
    default_time_format = '%Y-%m-%d %H:%M:%S'
    default_msec_format = '%s,%03d'
```
Since LogRender is "hidden" that doesn't feel like the right approach so instead I added init arguments to Console and RichHandler.

Update to Console's doc string:

             log_time_format (str, optional): Log time format if ``log_time`` is enabled. Defaults to "[%X] ".
    +        log_time_msec_format (str, optional): Log time msec format if ``log_time`` is enabled. Defaults to ``None``.
             highlighter (HighlighterType, optional): Default highlighter.

Update to RichHandlers's doc string:

             locals_max_string (int, optional): Maximum length of string before truncating, or None to disable. Defaults to 80.
    +        log_time_format (str, optional): Log time format if ``log_time`` is enabled. Defaults to "[%x %X] ".
    +        log_time_msec_format (str, optional): Log time msec format if ``log_time`` is enabled. Defaults to ``None``.

Note, I left the default log behaviour unchanged and if you want msecs you can do something like:
```handler = RichHandler(log_time_format='%H:%M:%S', log_time_msec_format='%s.%03d')```

**Reply by finnhughes** [reply]
I'd prefer to leverage the timestamp configuration in the Formatter class but assuming you want to maintain the current default time formatting for rich it's a bit tricky. I could add a "special" value of log_time_format like "default" or maybe the formatter object which would override the rich default to take the formatter default?
I could add a new argument called something like `log_time_format_from_formatter`

**Reply by willmcgugan** [reply]
I'd rather not add more configuration options. Just because there are quite a few already.

How about rather than a format string, I provide a `format_time` method on Console and LogHandler. You could extend that and format it in any way you like (you could even colorize it).

**Reply by finnhughes** [reply]
So pull formatting of time out of the hidden `LogRender` into `Console` and `RichHandler` and then leave it up to users like myself to override it if we want different formatting? With a tiny bit of refactoring `RichHandler` and `Console` can just use a shared function in `LogRender` for the default implementation?

If that's the way you'd prefer it was handled I can work up a PR for that in the coming days?

**Reply by willmcgugan** [reply]
That was pretty much it, but thinking about it some more, it still doesn't feel like an elegant solution.

Here's yet another alternative. Perhaps `time_format` can be a union of a str and a callable which takes a datetime and returns a Text instance.

If `time_format` is a str then use strftime as before, otherwise call it to get the Text for the time column. So you could implement your milliseconds format with something like the following:

```python
def render_time(log_time:datetime) -> Text:
    time = format_with_milliseconds(log_time)
    return Text(time)

console = Console(time_format=render_time)
```

Thoughts?

**Reply by finnhughes** [reply]
Thanks, I think that works, note in RichHandler we need to add ms to log_time e.g:
```python
        log_time = datetime.fromtimestamp(record.created) + timedelta(microseconds=1000 * record.msecs)
```
I can update my pull request that way, might be a day or two. Note hadn't thought of adding milliseconds to log_time until your message above and I think it makes the code much cleaner.

**Reply by finnhughes** [reply]
Presumably that means adding a `time_format` argument to `RichHandler` to mirror the one in `Console`?

**Reply by willmcgugan** [reply]
>   log_time = datetime.fromtimestamp(record.created) + timedelta(microseconds=1000 * record.msecs)

Do you need to do this? Docs say `created` is a float as returned from `time.time()` which means it should contain full resolution.

> Presumably that means adding a time_format argument to RichHandler to mirror the one in Console?

Makes sense.

**Reply by finnhughes** [reply]
Good point I don't think I need to, it was just `logging` handling milliseconds separately that confused me.

I think it's easier to handle  in `_log_render.py` if the callable just returns a `str` rather than `Text`. It lets `_log_render.py` handle the conversion to `Text` and not duplicating timestamps in the way it already does?

**Reply by finnhughes** [reply]
This PR looks a lot cleaner though trying to describe the two options for `log_time_format` in one line is a bit tricky.

**Reply by finnhughes** [reply]
Thanks again for reviews and suggestions, if the current PR isn't it, it should be very close.

Your really helpful suggestions mean it looks a lot better now 👏
